addon/@ember/test-waiters/wait-for.ts

```diff
@@ -38,21 +45,51 @@ type DecoratorArguments = [object, string, PropertyDescriptor, string?];
  *   }
  * }
  *
+ * @example
+ *
+ * import Component from '@ember/component';
+ * import { task } from 'ember-concurrency';
+ * import { waitFor } from 'ember-test-waiters';
+ *
+ * export default Component.extend({
+ *   doTaskStuff: task(waitFor(function* doTaskStuff() {
+ *     yield somethingAsync();
+ *   }
+ * });
+ *
+ * @example
+ *
+ * import Component from '@ember/component';
+ * import { task } from 'ember-concurrency-decorators';
+ * import { waitFor } from 'ember-test-waiters';
+ *
+ * export default class Friendz extends Component {
+ *   @task
+ *   @waitFor
+ *   *doTaskStuff() {
+ *     yield somethingAsync();
+ *   }
+ * }
+ *
  */
 export default function waitFor(fn: AsyncFunction<any[], any>, label?: string): Function;
+export default function waitFor(
+  fn: TaskFunction<any[], any>,
+  label?: string
+): TaskFunction<any[], any>;
 export default function waitFor(
   target: object,
   _key: string,
   descriptor: PropertyDescriptor,
   label?: string
 ): PropertyDescriptor;
 export default function waitFor(
-  ...args: AsyncFunctionArguments | DecoratorArguments
-): PropertyDescriptor | Function {
-  let isAsyncFunction = args.length < 3;
+  ...args: FunctionArguments | DecoratorArguments
+): PropertyDescriptor | Function | TaskFunction<any[], any> {
+  let isFunction = args.length < 3;
 
-  if (isAsyncFunction) {
-    let [fn, label] = args as AsyncFunctionArguments;
+  if (isFunction) {
```

Comment:
I disagree. I would argue that `isAsyncFunction` was actually a misnomer. `args.length < 3` tells us nothing about what kind of thing this function is wrapping, it only tells us how this function has been invoked -- as a function call or as a decorator.

At this point we need to make the decision of *how* to apply `wrapFunction()` to wrap the thing the caller wants us to wrap, not *what* the thing is the caller wants us to wrap. It's the function that `wrapFunction()` generates that determines *what* the thing is because we can't possibly distinguish between an async function, a non-async function, and a generator function until we actually call it and inspect the return value, which we can't do here.

So at this point we can really only have two cases, wait-for-was-invoked-as-a-function-call and wait-for-was-invoked-as-a-decorator, but if you like, I could clarify by renaming it to `isFunctionalInvocation` or something like that.

Reply:
Ya I think my relevant point is that in the prior implementation there were only two distinguishable paths: async functions and decorators. With this addition it's true that inspecting the length of the arguments isn't enough to unequivocally determine this, but my point still stands. I think the current version is less readable, and as such may be less maintainable. Is there anything we can do to address this?

Reply:
No, there's nothing we can do. In this function, the only way to be able to distinguish three paths instead of two would be if we could distinguish a generator function from an async function without calling that function, and that's not possible. Moreover, I will argue again that it's not now and never was thins function's job to do that.

`async functions and decorators` were not the two cases in the prior implementation. There were actually four cases:

1. Called in functional form, passed an async function
2. Called in function form, passes a "sync" function
3. Called in decorator form, passed an async function
4. Called in decorator form, passed a "sync" function

`async function` and `decorator` are not equivalent paths -- `called as a function` and `called as a decorator` are equivalent paths, and `called on an async function` and `called on a regular function` (and now `called on a generator function`) are equivalent paths. This function's job is to distinguish between `called as a function` and `called as a decorator` so it knows *how* to wrap the function it was passed, and the wrapper function returned from `wrapFunction()` is the one whose job it is to distinguish between `called on an async function` and `called on a regular function` (and now `called on a generator function`).

That's why I say `isAsyncFunction` was a misnomer -- it was incorrectly conflating how `waitFor()` was invoked (function or decorator) with what was passed to `waitFor()` (async function or regular function).

Make sense?

Reply:
Yep, makes perfect sense.